gravitee-am-factor/gravitee-am-factor-sms/src/main/java/io/gravitee/am/factor/sms/provider/SMSFactorProvider.java

```diff
@@ -0,0 +1,124 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.gravitee.am.factor.sms.provider;
+
+import com.google.i18n.phonenumbers.NumberParseException;
+import com.google.i18n.phonenumbers.PhoneNumberUtil;
+import com.google.i18n.phonenumbers.Phonenumber;
+import io.gravitee.am.factor.api.Enrollment;
+import io.gravitee.am.factor.api.FactorContext;
+import io.gravitee.am.factor.api.FactorProvider;
+import io.gravitee.am.factor.sms.SMSFactorConfiguration;
+import io.gravitee.am.gateway.handler.resource.ResourceManager;
+import io.gravitee.am.model.factor.EnrolledFactor;
+import io.gravitee.am.model.factor.EnrolledFactorSecurity;
+import io.gravitee.am.repository.exceptions.TechnicalException;
+import io.gravitee.am.resource.api.ResourceProvider;
+import io.gravitee.am.resource.api.mfa.MFAChallenge;
+import io.gravitee.am.resource.api.mfa.MFALink;
+import io.gravitee.am.resource.api.mfa.MFAResourceProvider;
+import io.gravitee.am.resource.api.mfa.MFAType;
+import io.reactivex.Completable;
+import io.reactivex.Single;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+
+import java.util.Locale;
+
+/**
+ * @author Eric LELEU (eric.leleu at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+public class SMSFactorProvider implements FactorProvider {
+
+    private static final Logger logger = LoggerFactory.getLogger(SMSFactorProvider.class);
+
+    @Autowired
+    private SMSFactorConfiguration configuration;
+
+    @Override
+    public Completable verify(FactorContext context) {
+        final String code = context.getData(FactorContext.KEY_CODE, String.class);
+        final EnrolledFactor enrolledFactor = context.getData(FactorContext.KEY_ENROLLED_FACTOR, EnrolledFactor.class);
+        ResourceManager component = context.getComponent(ResourceManager.class);
+        Single<ResourceProvider> provider = component.getResourceProvider(configuration.getGraviteeResource());
+        return provider.map(p -> {
+            if (p instanceof MFAResourceProvider) {
+                MFAResourceProvider mfaProvider = (MFAResourceProvider)p;
+                MFAChallenge challenge = new MFAChallenge(enrolledFactor.getSecurity().getValue(), code);
+                mfaProvider.verify(challenge);
+                return mfaProvider;
+            } else {
+                throw new TechnicalException("Resource referenced can't be used for MultiFactor Authentication  with type SMS");
+            }
+        }).ignoreElement();
```

Comment:
Kind of weird ... Maybe easier to move verify method from void to Completable ?

Reply:
Done